internal/filesystem/cgofuse/lock/lock.go

```diff
@@ -0,0 +1,196 @@
+package lock
+
+import (
+	"strings"
+	"sync"
+)
+
+// TODO: [review] This implementation is both incomplete
+// and likely suboptimal. Its current state is intended to be enough to be usable.
+// But it should be made optimal eventually, since it's the hottest code-path in FUSE.
+// (Every fuse operation must call these methods.)
+//
+// We also might want to move this pkg somewhere else.
+// It's fine under `filesystem/cgofuse` for now, but might be applicable to other pkgs.
+// We'd need to name things better to move it up to `/filesystem/something-lock`
+
+type (
+	pathLock struct {
+		pathMu,
+		dataMu sync.RWMutex
+	}
+	pathLockReference struct {
+		pathLock
+		// TODO: [review] Double check the logic.
+		// I'm pretty sure this count doesn't have to be atomic;
+		// when used within [pathLocker]. Since that has a global mutex.
+		referenceCount int
+	}
+	pathLockerMap map[string]*pathLockReference
+	// PathLocker is a hierarchical path locker,
+	// inspired by Ritik Malhotra's paper on path-based locks.
+	PathLocker struct {
+		lockTableMu sync.Mutex
+		lockTable   pathLockerMap
+	}
+	UnlockFunc = func()
+	lockFunc   = func()
+	// TODO: [Ame] English and names.
+	// sequenceFunc is an abstraction to reduce duplication.
+	// Intermediate components of a path, all take the same locks.
+	// The final component will have a unique (un)lock sequence
+	// that is dependant on the operation.
+	// This function may use the provided locks to generate that sequence.
+	sequenceFunc func(path, data *sync.RWMutex) (lockFunc, UnlockFunc)
+)
+
+func (lm pathLockerMap) upsert(path string) *pathLockReference {
+	if lock, ok := lm[path]; ok {
+		lock.referenceCount++
+		return lock
+	}
+	lock := &pathLockReference{referenceCount: 1}
+	lm[path] = lock
+	return lock
+}
+
+// TODO: [Ame] Does this English and notation make sense? Could it be more succinct?
+//
+// componentIndex returns indices that can be used
+// to retrieve components of a slash delimited string.
+//
+// The initial delimiter and final component are included.
+//
+// E.g. path `"/a/b/c"` produces indices `[]int{1, 2, 3, 6}`
+// path[:indices[0]] == "/"
+// path[:indices[1]] == "/a"
+// path[:indices[2]] == "/a/"
+// path[:indices[3]] == "/a/b/c"
+func componentIndex(path string) []int {
```

Comment:
This comment is true, but the logic/usage seems wrong.

There was a reason why the original operations-lock implementation (not the one in the diff) sliced paths this way. But the way paths are passed around, and the structures used for storing the mutexes - have changed. So it's not really relevant anymore.

As-is we're creating redundant locks for intermediate components.
And I think this diff broke the splitting logic too.

I'm going to fix and simplify this later.